### PR TITLE
Prevent signed integer overflow in `check_scan_over_group`

### PR DESCRIPTION
**Pull request by 0x12CC**
Check that the maximum value of `T` is large enough to store the results of the scan functions. This check prevents overflow for small types. Also, remove the largest `check_scan` test case.

### CODE REVIEW

File: tests/group_functions/group_scan.h [hunk 1]
```diff
@@ -303,6 +303,10 @@ void check_scan_over_group(sycl::queue& queue, sycl::range<D> range, OpT op,
   bool ret_type_e = false;
   bool ret_type_i = false;
 
+  if (std::sqrt(std::numeric_limits<T>::max()) + T(init) <= range.size()) {
+    return;
+  }
+
```
Comment (bader) [inline tests/group_functions/group_scan.h:309]: I don't think it's a good idea silently skipping all checks in this case.
Although this sounds like a test issue, rather than a bug in the tested implementation, we probably can use REQUIRE macro to "Check that the maximum value of T is large enough to store the results of the scan functions.".

Comment (bader) [reply tests/group_functions/group_scan.h:309]: Why do we launch the configuration causing the overflow? Isn't this a bug?

Comment (bader) [reply tests/group_functions/group_scan.h:309]: Right. This where I think we should change the logic to avoid an overflow. 
work_goup_size here is the maximum possible # of work-items the device can execute. Do we really need to the maximum # of work-items to test the scan functionality? If so, the input buffer must be initialized with zeros to avoid overflow.
But I don't think we need to run so many work-items, it's a good stress test, but an overkill for this test purpose.

https://github.com/KhronosGroup/SYCL-CTS/blob/47a05179ba09a364c4db03ce07fc387564eebdac/tests/group_functions/group_scan.h#L68 - this seems to be the right place to make sure that we can run the test w/o an overflow. BTW, it's better to use [the sum of arithmetic progression](https://en.wikipedia.org/wiki/Arithmetic_progression#Sum) instead of sqrt for the overflow check.

File: tests/group_functions/group_scan.h [hunk 2]
```diff
@@ -238,7 +238,7 @@ struct init_joint_scan_group {
 
       size_t work_group_size = work_group_range.size();
 
-      const size_t sizes[3] = {5, work_group_size / 2, 3 * work_group_size};
+      const size_t sizes[3] = {5, work_group_size / 2};
```
Comment (bader) [inline tests/group_functions/group_scan.h:241]: Either declare 2-element array or explicitly initialize the 3rd element.

Why the second element is `work_group_size / 2`? Why can't we put 1 or 2 instead (+ check that HW is able to run the work of that size)?

File: tests/group_functions/group_scan.h [hunk 3]
```diff
@@ -303,6 +303,11 @@ void check_scan_over_group(sycl::queue& queue, sycl::range<D> range, OpT op,
   bool ret_type_e = false;
   bool ret_type_i = false;
 
+  if (((range_size * (range_size + 1) / 2) + T(init)) >
+      std::numeric_limits<T>::max()) {
+    return;
+  }
+
```
Comment (bader) [inline tests/group_functions/group_scan.h:310]: I still think that silently skipping checks is a bad idea. Please, replace with an assertion.

**Review by AerialMantis — APPROVED** [review]

**Comment by fraggamuffin** [comment]
Ivan, please have a look. Thx.

**Comment by bader** [comment]
I'll wait for the feedback from @Melirius to merge.

**Review by keryell — APPROVED** [review]
Thanks.

**Comment by fraggamuffin** [comment]
Will wait one more week.

**Comment by bader** [comment]
Friendly reminder for @Melirius.

**Comment by bader** [comment]
Friendly reminder №2 for @Melirius.

@0x12CC, could you resolve merge conflicts with SYCL-2020 branch, please?

File: tests/group_functions/group_scan.h (hunk 2 above)
Comment (keryell) [reply tests/group_functions/group_scan.h]: That means which should even more comment the code.

**Comment by Melirius** [comment]
> Friendly reminder №2 for @Melirius.

Sorry, I'm at vacations now, be back on the next week.

**Comment by bader** [comment]
> > Friendly reminder №2 for @Melirius.
> 
> Sorry, I'm at vacations now, be back on the next week.

@Melirius, friendly reminder №3.

We have required approvals, so it's ready to be merged. I'm going to merge this PR next Wednesday if there are no additional requests for changes.
@Melirius, feel free provide your feedback regardless of the merge status. I'll make sure that we address post-merge review comments too.
